homeassistant/components/upnp/sensor.py

```diff
@@ -1,275 +1,256 @@
+        """Return the unit of measurement of this entity, if any."""
+        return self._sensor_type["unit"]
 
     @property
-    def icon(self) -> str:
-        """Icon to use in the frontend, if any."""
-        return "mdi:server-network"
+    def device_info(self) -> Mapping[str, any]:
+        """Get device info."""
+        return {
+            "connections": {(dr.CONNECTION_UPNP, self._device.udn)},
+            "name": self._device.name,
+            "manufacturer": self._device.manufacturer,
+            "model": self._device.model_name,
+        }
 
-    @property
-    def unit_of_measurement(self) -> str:
-        """Return the unit of measurement of this entity, if any."""
-        return self._type["unit"]
+    async def async_added_to_hass(self) -> None:
+        """Subscribe to sensors events."""
+        self._coordinator.async_add_listener(self.async_write_ha_state)
 
-    @Throttle(MIN_TIME_BETWEEN_UPDATES)
-    async def async_update(self):
-        """Get the latest information from the IGD."""
-        if self._type_name == BYTES_RECEIVED:
-            self._state = await self._device.async_get_total_bytes_received()
-        elif self._type_name == BYTES_SENT:
-            self._state = await self._device.async_get_total_bytes_sent()
-        elif self._type_name == PACKETS_RECEIVED:
-            self._state = await self._device.async_get_total_packets_received()
-        elif self._type_name == PACKETS_SENT:
-            self._state = await self._device.async_get_total_packets_sent()
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, SIGNAL_REMOVE_DEVICE, self._async_upnp_remove_sensor
+            )
+        )
 
+    async def _async_upnp_remove_sensor(self, device) -> None:
+        """Remove sensor."""
+        if self._device != device:
+            # not for us
+            return
 
-class PerSecondUPnPIGDSensor(UpnpSensor):
-    """Abstract representation of a X Sent/Received per second sensor."""
+        _LOGGER.debug("Removing sensor: %s", self.unique_id)
+        await self.async_remove()
 
-    def __init__(self, device, direction):
-        """Initialize sensor."""
-        super().__init__(device)
-        self._direction = direction
+    async def async_will_remove_from_hass(self) -> None:
+        """When entity will be removed from hass."""
+        self._coordinator.async_remove_listener(self.async_write_ha_state)
 
-        self._state = None
-        self._last_value = None
-        self._last_update_time = None
+
+class RawUpnpSensor(UpnpSensor):
+    """Representation of a UPnP/IGD sensor."""
 
     @property
-    def unit(self) -> str:
-        """Get unit we are measuring in."""
-        raise NotImplementedError()
+    def state(self) -> str:
+        """Return the state of the device."""
+        device_value_key = self._sensor_type["device_value_key"]
+        value = self._coordinator.data[device_value_key]
```

Comment:
Since you are returning exceptions, this can be an exception ! 

If it's an exception, available property should return False. isinstance(self._coordinator.data[device_value_key], Exception)`

Reply:
I expected the `DataUpdateCoordinator` to set `last_update_success` to false in case any of the requests failed. Anyway, fixed.

Reply:
It is not set to False, because by passing `return_exceptions=True`, the exceptions are returned as if it is valid data. `last_update_success` is only set to `False` if an exception was raised: aiohttp exceptions, TimeoutException or UpdateFailed exception.

Reply:
Generally if you return exceptions from a gather, you want to make sure you check the results and raise accordingly. You don't want to return them as if it's data.